OFDM Cyclic prefixer is now safer and more flexible.

m3x1m0m opened the pull request:
Made a new branch as I messed it up with the old one. @willcode I implemented your suggestion. If the way I solved it now is OK, I will change the other examples, as this commit breaks backward compatibility. The idea with a vector for different CPs required a new signature. However, the tx_ofdm.grc I fixed to demonstrate, that my code is working. New test cases are also provided along with the code.

drmpeg commented:
The cyclic prefixer is used extensively in the gr-dtv component. You need to update the following example flow graphs in gr-dtv/examples

dvbt_tx_2k.grc
dvbt_tx_8k.grc
germany-g10.grc
germany-g1.grc
germany-g2.grc
germany-g3.grc
germany-g4.grc
germany-g5.grc
germany-g6.grc
germany-g7.grc
germany-g8.grc
germany-g9.grc
vv001-cr35.grc
vv003-cr23.grc
vv004-8kfft.grc
vv005-8kfft.grc
vv007-16kfft.grc
vv008-16kfft.grc
vv009-4kfft.grc
vv010-2kfft.grc
vv011-1kfft.grc
vv012-64qam45.grc
vv013-64qam56.grc
vv014-64qam34.grc
vv015-8kfft.grc
vv016-256qam34.grc
vv017-paprtr.grc
vv018-miso.grc
vv019-norot.grc
vv034-dtg016.grc
vv035-dtg052.grc
vv036-dtg091.grc

drmpeg commented:
Just FYI. In the past, changes that modified the block API were not allowed in sub releases (3.7.x). A change like this would be put off for a major release (3.8).

m3x1m0m commented:
@drmpeg Thank you for bringing this up. I will wait for more comments. Also sleep one night over the code. If there is no more change in the interface I will look at the examples.

m3x1m0m commented:
Well, 3.8 is planned to be released soon, right?

drmpeg commented:
I believe the target for 3.8 is September. Not sure if that's "soon".

m3x1m0m commented:
What about new blocks? When it comes to OFDM I think some functionalities might be an enrichment e.g. a Carrier Deallocator or a CP Remover.

drmpeg commented:
New blocks are fine. There's no API to break.

drmpeg commented:
BTW, since there's a new project maintainer now, I'll let @marcusmueller comment and clarify on the API policy.

mbr0wn commented on gr-digital/include/gnuradio/digital/ofdm_cyclic_prefixer.h:
```diff
@@ -45,28 +45,37 @@ namespace gr {
+      static sptr make
+      (
+        int fft_len,
+        std::vector<int> cp_lengths,
```
This is a pretty massive API change. I think we can make it a much less problematic change if you add another, overloaded `make()` call with the old signature. We have that elsewhere.

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
 
     ofdm_cyclic_prefixer::sptr
-    ofdm_cyclic_prefixer::make(size_t input_size, size_t output_size, int rolloff_len, const std::string &len_tag_key)
+    ofdm_cyclic_prefixer::make(int fft_len, std::vector<int> cp_lengths, int rolloff_len, const std::string &len_tag_key)
```
Your overloaded make() call could simply create a vector of length 1 and then we don't need to overload the ctors, too.

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
-      // Flank of length 1 would just be rectangular
-      if (d_rolloff_len == 1) {
-	d_rolloff_len = 0;
+      for(const int& cp_len : d_cp_lengths) {
```
Not sure if C++11 constructs are already available on this branch.

mbr0wn replied:
(they are)

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
-	  d_down_flank[i-1] = 0.5 * (1 + cos(M_PI * (rolloff_len-i)/rolloff_len - M_PI));
-	}
+      // Sanity
+      for(unsigned i=0; i<d_cp_lengths.size(); i++) {
```
Can you spell out what exactly the sanity checks are that you're going to do? The Doxygen for this block would be a good place.

mbr0wn commented on gr-digital/include/gnuradio/digital/ofdm_cyclic_prefixer.h:
```diff
@@ -45,28 +45,37 @@ namespace gr {
+     * Different CP lengths as for instance needed in LTE are supported. This
+     * is why one of the inputs is std::vector<int>. After every CP given has
+     * been prepended to symbols, each with the length of the IFFT operation,
+     * the mechanism will wrap arround and start over. To give an example, the
```
typo "around)

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
+          break;
+        }
+        if (i == d_cp_lengths.size()-1) {
+          throw std::invalid_argument(this->alias() + std::string(": It does not make sense to only have one or a couple of CPs with length 0."));
```
Prefer a positive error message ("You must provide X") instead of negative ("What you provided is invalid") so it's clear, based on the error message, what the user has to change.

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
-	set_output_multiple(d_output_size);
-      } else {
-	set_tag_propagation_policy(TPP_DONT);
+      if (d_cp_lengths.empty()) {
```
Please move this check above the previous. It doesn't make a huge difference, but it's a good practice to do these checks in the order in which they fail fastest.

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
-      } else {
-	set_tag_propagation_policy(TPP_DONT);
+      if (d_cp_lengths.empty()) {
+        throw std::invalid_argument(this->alias() + std::string(": CP lengths vector can not be empty. Please enter at least (d,) (a single Python tuple) with d being a integer CP length."));
```
This message assumes the usage of Python. which we can't assume.

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
+      set_relative_rate(d_cp_max + d_fft_len);
+      // Flank of length 1 would just be rectangular.
+      if(d_rolloff_len == 1) {
+        throw std::invalid_argument(this->alias() + std::string(": A rolloff length of 1 does not make sense, as it results in a boxcar function."));
```
This is another API change. I think I made the previous version because there are cases when people think they need to provide 1 when what they mean is 0. I would strongly advocate leaving the old behaviour unchanged, but add a warning-level log message.

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
+        // More sanity
+        for(const int& cp_len : d_cp_lengths) {
+          if(d_rolloff_len > cp_len) {
+            throw std::invalid_argument(this->alias() + std::string(": Rolloff length must be smaller than the cyclic prefix."));
```
This error message can be improved, since you have multiple cyclic prefixes. Maybe: "Rolloff length must be smaller than any of the cyclic prefixes.".

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
+      if(d_len_tag_key.empty()) {
+        // noutput_items is set to be a multiple of the largest possible output size.
+        // It is always OK to return less (in case of the shorter CP).
+        set_output_multiple(d_cp_max);
```
If you have d_cp_max, you might as well also collect d_cp_min for some of the sanity checks above. Saves you a loop.

mbr0wn commented on gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:
```diff
@@ -31,135 +31,147 @@ namespace gr {
+        if(d_rolloff_len) {
+          for(int i = 0; i < d_rolloff_len-1; i++) {
+            out[i] = out[i] * d_up_flank[i] + d_delay_line[i];
+            /* This basically a cyclic suffix, but completely shifted into the next symbol.
```
typo: "This *is* basically..."

m3x1m0m commented:
@mbr0wn Thank you very much for your comments. I removed the C++11 loop, improved documentation, overloaded the make function, as you suggested, and added a test case to check if the old interface still behaves as before. Now it is not necessary to change any old examples or other use cases of this block!

drmpeg commented:
Doesn't build here on Ubuntu 16.04 and gcc 5.4. Here's the error.

```/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc: In static member function ‘static gr::digital::ofdm_cyclic_prefixer::sptr gr::digital::ofdm_cyclic_prefixer::make(size_t, size_t, int, const string&)’:
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:38:80: error: in C++98 ‘cp_lengths’ must be initialized by constructor, not by ‘{...}’
       std::vector<int> cp_lengths = {static_cast<int>(output_size - input_size)}; // Cast to silence compiler :(
                                                                                ^
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:38:80: error: converting to ‘std::vector<int>’ from initializer list would use explicit constructor ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>::size_type, const value_type&, const allocator_type&) [with _Tp = int; _Alloc = std::allocator<int>; std::vector<_Tp, _Alloc>::size_type = long unsigned int; std::vector<_Tp, _Alloc>::value_type = int; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<int>]’
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:38:80: warning: narrowing conversion of ‘(int)(((unsigned int)output_size) - ((unsigned int)input_size))’ from ‘int’ to ‘std::vector<int>::size_type {aka long unsigned int}’ inside { } is ill-formed in C++11 [-Wnarrowing]
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc: In constructor ‘gr::digital::ofdm_cyclic_prefixer_impl::ofdm_cyclic_prefixer_impl(int, std::vector<int>, int, const string&)’:
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:95:33: warning: range-based ‘for’ loops only available with -std=c++11 or -std=gnu++11
         for(const int& cp_len : d_cp_lengths) {
                                 ^
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc: In member function ‘virtual int gr::digital::ofdm_cyclic_prefixer_impl::work(int, gr_vector_int&, gr_vector_const_void_star&, gr_vector_void_star&)’:
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:165:9: warning: ‘auto’ changes meaning in C++11; please remove it [-Wc++0x-compat]
         auto last_state = d_state > 0 ? d_state-1 : d_cp_lengths.size()-1;
         ^
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:165:14: error: ‘last_state’ does not name a type
         auto last_state = d_state > 0 ? d_state-1 : d_cp_lengths.size()-1;
              ^
/media/re/ssd/gnuradio/gr-digital/lib/ofdm_cyclic_prefixer_impl.cc:169:92: error: ‘last_state’ was not declared in this scope
           tags[i].offset = ( (tags[i].offset - nitems_read(0)) * (d_fft_len + d_cp_lengths[last_state]) ) + nitems_written(0);
                                                                                            ^
gr-digital/lib/CMakeFiles/gnuradio-digital.dir/build.make:1426: recipe for target 'gr-digital/lib/CMakeFiles/gnuradio-digital.dir/ofdm_cyclic_prefixer_impl.cc.o' failed
make[2]: *** [gr-digital/lib/CMakeFiles/gnuradio-digital.dir/ofdm_cyclic_prefixer_impl.cc.o] Error 1
```

m3x1m0m commented:
C++98, is this not considered outdated? I changed a few newer language constructs I am making use of and am compiling it right now on my machine.

mbr0wn commented:
@m3x1m0m C++98 *is* outdated, no doubt about it, but we can't just go ahead and change the compiler settings within a release cycle. Like @noc0lour says, on 3.8, we've already moved to C++11. Maybe we just apply it there...

drmpeg commented:
By changing the key names in digital_ofdm_cyclic_prefixer.xml, you've made it not backwards compatible. You can either go back to the old key names or regenerate the 32 flow graphs in gr-dtv/examples I listed above. Going back to the original key names seems significantly easier.

Otherwise, the block works fine for OFDM in gr-dtv.

m3x1m0m commented:
@mbr0wn I am a bit confused. I think I remember, that in the last dev call it was said, that the maint-3.7 branch is where the current release is maintained and in the master branch is where new hot features belong. Is this not correct? Where would I put new stuff, which should explicitly end up in the next release?

@drmpeg Yes, you are totally right. I will look into this again.

m3x1m0m commented:
@drmpeg I changed all the examples you listed. Please let me know if I forgot something or made a mistake.

drmpeg commented:
Looks good.

marcusmueller commented:
This uses C++11, which we'll have only after the next merge, so merging right after that :)

marcusmueller commented:
merge test: #1987

dkozel commented:
`master` now has C++11 as a minimum version dependency FYI.

m3x1m0m commented:
I briefly looked into it. It is on my todo list. I will fix it asap.

noc0lour commented:
Dang, the flowgraph conversion to YAML messed with this PR (and needs to be accounted for).

mbr0wn commented:
@m3x1m0m this has taken way too long on our part. I'll take the branch, rebase, and resubmit. I still like this change.

mbr0wn commented:
Continued here: https://github.com/gnuradio/gnuradio/pull/3043
